File: testdata/precision-suite-php/clean_safe_db.php
<?php
// Safe database idioms — every one is the correct, guarded form, so a precise
// scanner fires nothing. Zero findings expected.

// A prepared statement with a bound placeholder: the driver binds the value, so
// no injection is possible. The tainted value never touches the SQL string.
function findUser($pdo)
{
    $id = $_GET['id'];
    $stmt = $pdo->prepare("SELECT name, email FROM users WHERE id = :id");
    $stmt->execute([':id' => $id]);
    return $stmt->fetch();
}

// Numeric coercion via intval removes every injection metacharacter before the
// value reaches the query.
function findByPage($conn)
{
    $raw = $_GET['page'];
    $page = intval($raw);
    return mysqli_query($conn, "SELECT * FROM posts LIMIT " . $page);
}

File: testdata/precision-suite/clean_safe_db.go
// Safe database and command idioms in Go — every one is the correct, guarded
// form, so a precise scanner fires nothing. Zero findings expected.
package store

import (
	"database/sql"
	"os/exec"
)

// lookup uses a parameterized query ($1 placeholder) — the driver binds the
// value, so no injection is possible.
func lookup(db *sql.DB, id string) (*sql.Rows, error) {
	return db.Query("SELECT name, email FROM users WHERE id = $1", id)
}

// listDir runs a fixed argument vector (no shell), with the user value passed
// as a distinct argument rather than interpolated into a command string.
func listDir(dir string) ([]byte, error) {
	return exec.Command("ls", "-la", "--", dir).Output()
}

// allowedRegions is a fixed allowlist; only values it contains reach the callee.
var allowedRegions = map[string]bool{"us-east-1": true, "eu-west-1": true}

// region validates input against the allowlist before use.
func region(candidate string) string {
	if allowedRegions[candidate] {
		return candidate
	}
	return "us-east-1"
}
